cmd/admin-handlers.go

```diff
@@ -108,93 +121,49 @@ func (adminAPI adminAPIHandlers) ServiceStatusHandler(w http.ResponseWriter, r *
 	writeSuccessResponseJSON(w, jsonBytes)
 }
 
-// ServiceRestartHandler - POST /?service
-// HTTP header x-minio-operation: restart
-// ----------
-// Restarts minio server gracefully. In a distributed setup,  restarts
-// all the servers in the cluster.
-func (adminAPI adminAPIHandlers) ServiceRestartHandler(w http.ResponseWriter, r *http.Request) {
-	adminAPIErr := checkRequestAuthType(r, "", "", "")
-	if adminAPIErr != ErrNone {
-		writeErrorResponse(w, adminAPIErr, r.URL)
-		return
-	}
-
-	// Reply to the client before restarting minio server.
-	writeSuccessResponseHeadersOnly(w)
-
-	sendServiceCmd(globalAdminPeers, serviceRestart)
-}
-
-// setCredsReq request
-type setCredsReq struct {
-	Username string `xml:"username"`
-	Password string `xml:"password"`
-}
-
-// ServiceCredsHandler - POST /?service
-// HTTP header x-minio-operation: creds
+// ServiceStopNRestartHandler - POST /minio/admin/v1/service
+// Body: {"action": <restart-action>}
 // ----------
-// Update credentials in a minio server. In a distributed setup, update all the servers
-// in the cluster.
-func (adminAPI adminAPIHandlers) ServiceCredentialsHandler(w http.ResponseWriter, r *http.Request) {
-	// Authenticate request
+// Restarts/Stops minio server gracefully. In a distributed setup,
+// restarts all the servers in the cluster.
+func (adminAPI adminAPIHandlers) ServiceStopNRestartHandler(w http.ResponseWriter, r *http.Request) {
 	adminAPIErr := checkRequestAuthType(r, "", "", "")
 	if adminAPIErr != ErrNone {
 		writeErrorResponse(w, adminAPIErr, r.URL)
 		return
 	}
 
-	// Avoid setting new credentials when they are already passed
-	// by the environment.
-	if globalIsEnvCreds {
-		writeErrorResponse(w, ErrMethodNotAllowed, r.URL)
-		return
-	}
-
-	// Load request body
-	inputData, err := ioutil.ReadAll(r.Body)
+	body, err := ioutil.ReadAll(r.Body)
 	if err != nil {
-		writeErrorResponse(w, ErrInternalError, r.URL)
+		errorIf(err, "Failed to read request body")
+		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
 		return
 	}
 
-	// Unmarshal request body
-	var req setCredsReq
-	err = xml.Unmarshal(inputData, &req)
+	var sa madmin.ServiceAction
+	err = json.Unmarshal(body, &sa)
 	if err != nil {
-		errorIf(err, "Cannot unmarshal credentials request")
-		writeErrorResponse(w, ErrMalformedXML, r.URL)
+		writeErrorResponse(w, ErrMalformedPOSTRequest, r.URL)
+		errorIf(err, "Error parsing body JSON")
 		return
 	}
 
```

Comment:
more concise: `err = json.NewDecoder(r.Body).Decode(&sa)`

Reply:
Thanks! Fixed this in all except for PUT config as that does some extra checking for duplicate JSON keys.